### PR TITLE
Add `--no-rewrite` flag to `migrate import` command

### PR DESCRIPTION
**Pull request by PastelMobileSuit**
This PR adds a new `--no-rewrite` flag to the `migrate import` command. When this flag is given, `migrate import` takes a list of files as arguments and imports these files into Git LFS. It does this by rewriting the git tree in a new commit, without modifying git history.

Some of the code in this PR was based off of code previously written by @ttaylorr. Also, a big thank you to @ttaylorr for pairing with me on parts of this PR!

### CODE REVIEW

File: commands/command_migrate.go [hunk 1]
```diff
@@ -287,6 +294,8 @@ func init() {
 	importCmd := NewCommand("import", migrateImportCommand)
 	importCmd.Flags().BoolVar(&migrateVerbose, "verbose", false, "Verbose logging")
 	importCmd.Flags().StringVar(&objectMapFilePath, "object-map", "", "Object map file")
+	importCmd.Flags().BoolVar(&migrateNoRewrite, "no-rewrite", false, "Add new history without rewriting previous")
```
Comment (ttaylorr) [inline commands/command_migrate.go]: I think that these are great short descriptions of what the `--no-rewrite` and `--message` flags do, which is exactly what is appropriate for these.

Could you also add a longer explanation of what these flags are in [`git-lfs-migrate(1)`](https://github.com/git-lfs/git-lfs/blob/master/docs/man/git-lfs-migrate.1.ronn)? The `docs/man` tree is used to automatically generate `commands/mancontent.go`, and this in turn is what is shown when you run `git lfs migrate --help`, or `git lfs help migrate`, for instance.

The existing prose should be a good place to start in terms of what level of detail is appropriate.

File: commands/command_migrate_import.go [hunk 2]
```diff
@@ -29,6 +30,67 @@ func migrateImportCommand(cmd *cobra.Command, args []string) {
 	}
 	defer db.Close()
 
+	if migrateNoRewrite {
+		if len(args) == 0 {
+			ExitWithError(errors.Errorf("fatal: the migrate import command requires a list of files to import when using the --no-rewrite flag"))
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Great; I think dying here is the right thing to do. Let's shorten this message a little bit while we're here. For instance, the phrase "the migrate import command" might be unnecessary, since this will be the first line of output after a user runs `git lfs migrate import`.

What about: "fatal: expected one or more file(s) with --no-rewrite"?

File: commands/command_migrate_import.go [hunk 3]
```diff
@@ -29,6 +30,67 @@ func migrateImportCommand(cmd *cobra.Command, args []string) {
...
+		}
+
+		ref, err := git.CurrentRef()
+		if err != nil {
+			ExitWithError(errors.Wrap(err, "fatal: unable to find current reference"))
+		}
+
+		sha, _ := hex.DecodeString(ref.Sha)
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Great; throwing this error away is reasonable since we trust `git.CurrentRef()` to always give back a valid SHA-256 `ref.Sha`.

File: commands/command_migrate_import.go [hunk 4]
```diff
@@ -29,6 +30,67 @@ func migrateImportCommand(cmd *cobra.Command, args []string) {
...
+		commit, err := db.Commit(sha)
+		if err != nil {
+			ExitWithError(errors.Wrap(err, "fatal: unable to load commit"))
+		}
+
+		root := commit.TreeID
+
+		filter := git.GetAttributeFilter(cfg.LocalWorkingDir(), cfg.LocalGitDir())
+		if len(filter.Include()) == 0 && len(filter.Exclude()) == 0 {
+			ExitWithError(errors.Errorf("fatal: no git lfs filters setup in .gitattributes"))
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Let's use "found" instead of "setup".

File: commands/command_migrate_import.go [hunk 5]
```diff
@@ -29,6 +30,67 @@ func migrateImportCommand(cmd *cobra.Command, args []string) {
...
+		}
+
+		for _, file := range args {
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: I think one interesting consequence of `rewriteTree()` processing one file at a time is that there are some penalties that we'll pay if we get part of the way through this loop. In particular, if there are `m` arguments where `m` is large, and the `m-1`th of them doesn't match a pattern in `.gitattributes`, we'll have created a lot of tree objects that are dangling in the object database.

So, I think that we could run this loop twice, once to check that all given filenames match at least one pattern, and then once more to call `rewriteTree()`, which would mean we spend more time here. I think that's an OK tradeoff for not creating more objects than we need.

Of course, we could also look into ways to make `rewriteTree()` take more than one file and therefore create less dangling tree objects, but I'll leave it up to you if you want to go down that path.

File: commands/command_migrate_import.go [hunk 6]
```diff
@@ -29,6 +30,67 @@ func migrateImportCommand(cmd *cobra.Command, args []string) {
...
+
+		name, email := cfg.CurrentCommitter()
+		author := fmt.Sprintf("%s <%s>", name, email)
+
+		filelist := strings.Join(args, ",")
+		oid, err := db.WriteCommit(&odb.Commit{
+			Author:    author,
+			Committer: author,
+			ParentIDs: [][]byte{sha},
+			Message:   generateMigrateCommitMessage(cmd, &filelist),
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Let's inline `filelist`.

File: commands/command_migrate_import.go [hunk 7]
```diff
@@ -202,3 +283,98 @@ func trackedToBlob(db *odb.ObjectDatabase, patterns *tools.OrderedSet) ([]byte,
...
+	if err != nil {
+		return nil, err
+	}
+
+	splits := strings.SplitN(path, "/", 2)
+
+	switch len(splits) {
+	case 1:
+		// The path points to an entry at the root of this tree, so it must be a blob.
+		// Try to replace this blob with a git lfs pointer.
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Nit; casing on "Git LFS".

File: commands/command_migrate_import.go [hunk 8]
```diff
@@ -202,3 +283,98 @@ func trackedToBlob(db *odb.ObjectDatabase, patterns *tools.OrderedSet) ([]byte,
...
+		if _, err := clean(lfs.NewGitFilter(cfg), &buf, blob.Contents, blobEntry.Name, blob.Size); err != nil {
+			return nil, err
+		}
+
+		newBlob := odb.Blob{
+			Contents: &buf,
+			Size:     int64(buf.Len()),
+		}
+
+		newOid, err := db.WriteBlob(&newBlob)
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Nit, let's inline `newBlob`.

File: commands/command_migrate_import.go [hunk 9]
```diff
@@ -202,3 +283,98 @@ func trackedToBlob(db *odb.ObjectDatabase, patterns *tools.OrderedSet) ([]byte,
...
+
+		blobEntry := tree.Entries[index]
+		blob, err := db.Blob(blobEntry.Oid)
+		if err != nil {
+			return nil, err
+		}
+
+		var buf bytes.Buffer
+
+		if _, err := clean(lfs.NewGitFilter(cfg), &buf, blob.Contents, blobEntry.Name, blob.Size); err != nil {
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Why do we need to pass `lfs.NewGitFilter(cfg)`? I can't recall how the current implementation of `clean()` works, and whether or not we need this. If we do need it, can it be saved elsewhere instead of being constructed every time?

File: commands/command_migrate_import.go [hunk 10]
```diff
@@ -202,3 +283,98 @@ func trackedToBlob(db *odb.ObjectDatabase, patterns *tools.OrderedSet) ([]byte,
...
+		if err != nil {
+			return nil, err
+		}
+
+		newEntry := odb.TreeEntry{
+			Name:     splits[0],
+			Filemode: blobEntry.Filemode,
+			Oid:      newOid,
+		}
+		tree = tree.Merge(&newEntry)
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Nit; let's inline `newEntry` as well, to keep the current scope as narrow as possible.

File: commands/command_migrate_import.go [hunk 11]
```diff
@@ -202,3 +283,98 @@ func trackedToBlob(db *odb.ObjectDatabase, patterns *tools.OrderedSet) ([]byte,
...
+		})
+
+		return db.WriteTree(tree)
+
+	default:
+		return nil, errors.Errorf("error parsing path %s", path)
+	}
+}
+
+// findEntry searches a tree for the desired entry, and returns the
```
Comment (ttaylorr) [inline commands/command_migrate_import.go]: Nit; This line is wrapped at fewer than 80 characters.

File: git/attribs.go [hunk 12]
```diff
@@ -97,6 +98,22 @@ func GetAttributePaths(workingDir, gitDir string) []AttributePath {
...
+// configured with the filter=lfs attribute as a file path filter which
+// file paths can be matched against
+// workingDir is the root of the working copy
+// gitDir is the root of the git repo
+func GetAttributeFilter(workingDir, gitDir string) *filepathfilter.Filter {
+	paths := GetAttributePaths(workingDir, gitDir)
+	patterns := make([]filepathfilter.Pattern, len(paths))
+
+	for i, path := range paths {
+		patterns[i] = filepathfilter.NewPattern(path.Path)
```
Comment (ttaylorr) [inline git/attribs.go]: In case we want to change this later in a way that takes away the 1-1 correspondence between paths and patterns, we can minimize that diff by calling `append()` instead of using the slicing operator `[]`. In order to make that work as-expected, you'll want to change L108 to initialize `patterns` with a capacity of `len(paths)`, instead of a `len` of `len(paths)`.

File: test/test-migrate-fixtures.sh [hunk 13]
```diff
@@ -44,6 +44,46 @@ setup_local_branch_with_gitattrs() {
...
+# setup_local_branch_with_nested_gitattrs creates a repository as follows:
+#
+#   A---B
+#        \
+#         refs/heads/master
+#
+# - Commit 'A' has 120, in a.txt, and a corresponding entry in .gitattributes. There is also
+#   140 in a.md, with no corresponding entry in .gitattributes.
+#   It also has 140 in subtree/a.md, and a corresponding entry in subtree/.gitattributes
+setup_local_branch_with_nested_gitattrs() {
```
Comment (ttaylorr) [inline test/test-migrate-fixtures.sh]: ✨ Awesome!

File: test/test-migrate-import-no-rewrite.sh [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+  prev_commit_oid="$(git rev-parse HEAD)"
+
+  git lfs migrate import --no-rewrite *.txt
+
+  # Ensure our desired files were imported into git-lfs
+  assert_pointer "refs/heads/master" "a.txt" "$txt_oid" "120"
+  assert_local_object "$txt_oid" "120"
+
+  # Ensure the git history remained the same
+  new_commit_oid="$(git rev-parse HEAD~1)"
```
Comment (ttaylorr) [inline test/test-migrate-import-no-rewrite.sh]: ✨ Nice.

File: test/test-migrate-import-no-rewrite.sh [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+end_test
+
+begin_test "migrate import --no-rewrite (no .gitattributes)"
+(
+  set -e
+
+  setup_multiple_local_branches
+
+  # Ensure command fails if no .gitattributes files are present
+  [ !"$(git lfs migrate import --no-rewrite *.txt *.md)" ]
```
Comment (ttaylorr) [inline test/test-migrate-import-no-rewrite.sh]: Do you think we could check that the error message this invocation gives is what we expect it to be? I think one way to do this would be to write something of the ilk:

```bash
git lfs migrate import --no-rewrite *.txt *.md 2>&1 | tee migrate.log
if [ ${PIPESTATUS[0]} -eq 0 ]; then
  echo >&2 "fatal: expected git lfs migrate import --no-rewrite to fail, didn't"
fi
grep "..." migrate.log
```

File: commands/command_migrate_import.go (hunk 9 above)
Comment (PastelMobileSuit) [reply commands/command_migrate_import.go]: It looks like it's being used in a few places in the `clean()` implementation, but it shouldn't change throughout the course of the entire operation. I agree that it makes more sense to construct one in `migrateImportCommand()` and pass it down into `rewriteTree()`, so I'll go ahead and change that

File: commands/command_migrate_import.go (hunk 5 above)
Comment (PastelMobileSuit) [reply commands/command_migrate_import.go]: Ah okay, thanks! I didn't think about the possibility of dangling objects.

I agree with you on running the loop twice and think that's the best option, since we won't be able to avoid checking files against the `.gitattributes` filters.

**Comment by PastelMobileSuit** [comment]
@ttaylorr Thank you for the review! I've gone ahead and pushed up some new commits addressing your feedback

File: test/test-migrate-import-no-rewrite.sh [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    exit 1
+  fi
+
+  # Ensure the provided commit message was used
+  commit_msg="$(git log -1 --pretty=format:%s)"
+  if [ "$commit_msg" != "" ]; then
+    exit 1
+  fi
+)
+end_test
```
Comment (ttaylorr) [inline test/test-migrate-import-no-rewrite.sh]: Nit: missing trailing LF.

File: docs/man/git-lfs-migrate.1.ronn [hunk 17]
```diff
@@ -84,13 +84,39 @@ options and these additional ones:
...
     format is CSV with this pattern: `OLD-SHA`,`NEW-SHA`
 
-If `--include` or `--exclude` (`-I`, `-X`, respectively) are given, the
+* `--no-rewrite`
+    Migrate large objects to Git LFS in a new commit without rewriting git history.
+    Please note that when this option is used, the `migrate import` command will expect
+    a different argument list, specialized options will become available, and the core 
+    `migrate` options will be ignored. See [IMPORT (NO REWRITE)].
+
+If `--no-rewrite` is not provided and `--include` or `--exclude` (`-I`, `-X`, respectively) are given, the
```
Comment (ttaylorr) [inline docs/man/git-lfs-migrate.1.ronn]: Nit: it looks like a few lines in this file are longer than 80 columns. Can you make sure that the wrapping is set properly for `.ronn` files?

File: docs/man/git-lfs-migrate.1.ronn [hunk 18]
```diff
@@ -84,13 +84,39 @@ options and these additional ones:
...
+
+The `import` mode has a special sub-mode enabled by the `--no-rewrite` flag. This sub-mode
+will migrate large objects to pointers as in the base `import` mode, but will do so in a new commit
+without rewriting Git history. When using this sub-mode, the base `migrate` options, 
+such as `--include-ref`, will be ignored, as will those for the base `import` mode. 
+The `migrate` command will also take a different argument list.
+
+The `--no-rewrite` sub-mode supports the following options and arguments:
+
+* `-m` <message> `--message`=<message>
```
Comment (ttaylorr) [inline docs/man/git-lfs-migrate.1.ronn]: I think the backticks should wrap from the first `-` to the last `>` in each example.

File: docs/man/git-lfs-migrate.1.ronn [hunk 19]
```diff
@@ -84,13 +84,39 @@ options and these additional ones:
...
-If neither of those flags are given, the gitattributes will be incrementally
+If `--no-rewrite` is not provided and neither of those flags are given, the gitattributes will be incrementally
 modified to include new filepath extensions as they are rewritten in history.
 
+### IMPORT (NO REWRITE)
+
+The `import` mode has a special sub-mode enabled by the `--no-rewrite` flag. This sub-mode
+will migrate large objects to pointers as in the base `import` mode, but will do so in a new commit
+without rewriting Git history. When using this sub-mode, the base `migrate` options, 
+such as `--include-ref`, will be ignored, as will those for the base `import` mode.
```
Comment (ttaylorr) [inline docs/man/git-lfs-migrate.1.ronn]: I think that ignoring `--include-ref` is totally OK here, but we should perhaps add a note in the documentation that all interested branches will have to merge new commits generated by `git lfs migrate import --no-rewrite` if they _also_ want to take new commits in.

File: docs/man/git-lfs-migrate.1.ronn [hunk 20]
```diff
@@ -189,6 +215,18 @@ $ git lfs migrate import --everything --include="*.zip"
 
 Note: This will require a force push to any existing Git remotes.
 
+### Migrate without rewriting local history
+
+You can also migrate files without modifying the existing history of your respoitory:
+
+```
+# Without a specified commit message
```
Comment (ttaylorr) [inline docs/man/git-lfs-migrate.1.ronn]: Let's move these comments out of the code block, and replace the shell with 4 spaces, and then `git lfs migrate ...` (and without the `$`).

File: git/attribs.go [hunk 21]
```diff
@@ -73,7 +73,7 @@ func GetAttributePaths(workingDir, gitDir string) []AttributePath {
 				fields := strings.Fields(line)
 				pattern := fields[0]
 				if len(reldir) > 0 {
-					pattern = filepath.Join(reldir, pattern)
+					pattern = filepath.ToSlash(filepath.Join(reldir, pattern))
```
Comment (PastelMobileSuit) [inline git/attribs.go]: The necessity of this change was prompted by a consistent Windows test failure I was seeing on a VM. Every time I ran the nested `.gitattributes` test, I saw a failure.

It turns out the problem is that `filepath.Join()` uses `\` as the separator character on Windows, rather than `/`. As a result, a `*.txt` pattern in a folder `subtree` would be parsed as `subtree\*.txt`, and instead of acting as a path separator, the `\` would escape the `*` character, so the pattern would then only match a file with the name `subtree*.txt`.

**Comment by PastelMobileSuit** [comment]
Okay I've pushed up some additional commits to address feedback and have finally fixed all the AppVeyor test failures, so this should be ready to go!

File: git/attribs.go (hunk 21 above)
Comment (ttaylorr) [reply git/attribs.go]: Great fix! One nit is that `filepath.Join` calls `filepath.Clean` (see \[[1](https://github.com/golang/go/blob/master/src/path/filepath/path_unix.go#L45)\]), which can have unintended consequences with certain pattern types (i.e., escape sequences are removed, etc.)

How about we shorten this to:

```go
pattern = fmt.Sprintf("%s/%s", reldir, pattern)
```

or:

```go
pattern = strings.Join([]string{reldir, pattern}, "/")
```

Comment (PastelMobileSuit) [reply git/attribs.go]: @ttaylorr Hey! I believe this code was already being used as-is with the `track` command, so I think that was already removing escape characters from patterns. I did some investigation to confirm this, and `git lfs track "*\*.mov"` gets converted to `*/*.mov` in the `.gitattributes` file, and if you manually change it to `*\*.mov`, that's not matched by `track` on a file with the name `a*.mov`.

**Comment by hoeler** [comment]
@PastelMobileSuit @ttaylorr 
Hi!
I am not sure I grasp the use-case for the new "--no-rewrite" feature of git lfs import:
I was looking for a way to migrate our huge repository to lfs withouth destroying all of our (numerous) references to commit SHAs. I came to the conclusion that this is not possible and we will need to do a clean start with a rewritten history.
Then I found the announcement of the "--no-rewrite" feature and thought it may provide a way of moving all required files to lfs without changing the history (maybe by leaving the SHA intact although the content of the commits has changed). But this does not seem to be the use-case for this feature.

Can you please elaborate on when to use this feature? Sorry if I am missing something obvious, but the manual for migrate did not help me understand it and there is no further documentation that I can find.

Thanks a lot!

**Comment by ttaylorr** [comment]
Hi, @hoeler. Thanks for writing in, and I'd be more than happy to answer
your question. It seems to me that the central point is:

> Then I found the announcement of the "--no-rewrite" feature and
> thought it may provide a way of moving all required files to lfs
> without changing the history (maybe by leaving the SHA intact although
> the content of the commits has changed).

Which isn't quite the case. The '--no-rewrite' flag works by _adding_
new commits, instead of _rewriting_ existing ones.

The SHA-1 of a commit is dependent upon the contents of the files in
that commit's root tree, the parents, message, and etc., of the commit
itself. So, there is no way to "rewrite" the contents of a commit
without also changing its SHA-1.

'--no-rewrite' is used in a case where (1) you have files in Git that
should have been committed with Git LFS, but weren't, and (2) those
files are under 100 MiB (thus, don't prevent pushing to GitHub.com),
and/or (3) history rewriting is not an option.

Per (3), when you run 'git lfs migrate <import | export>', you have to
_also_ ensure that each clone of the repository has (1) fetched and (2)
reset their local references to match the remote's. When running 'git
lfs migrate import' with '--no-rewrite', the repository's cloners need
only run 'git pull' to receive the new changes.

The central difference between the two from the standpoint of "what are
my references?" is: 'git lfs migrate import' (with or without
'--no-rewrite') changes your references, the difference is whether or
not the old values are still valid locations in your repository's line
of history or not.

**Comment by hoeler** [comment]
Hi @ttaylorr, thank you for the explanation.
I am still not sure if i got it right though:

1) So basically "--no-rewrite" only works with the tip of a branch, by replacing existing files passed as arguments with LFS pointers and commiting that as a new change? Isn't this the same as manually removing the files from git and simply re-adding them under LFS control?

2) As I understand it now the "--no-rewrite" option does not help with getting rid of large files referenced somewhere in the git history, because this will _always_ require a rewrite of the history, correct?

**Comment by ttaylorr** [comment]
> 1) So basically "--no-rewrite" only works with the tip of a branch, by
> replacing existing files passed as arguments with LFS pointers and
> commiting that as a new change? Isn't this the same as manually
> removing the files from git and simply re-adding them under LFS
> control?

I think that the gist of that is correct, but 'git lfs migrate import
--no-rewrite' has two attributes that make it different than 'git rm
--cached', which are:

  1. Change each file in a separate commit, and

  2. Perform the change in Git LFS, so that it will work even if 'git
     lfs install' hasn't been run.

> 2) As I understand it now the "--no-rewrite" option does not help with
> getting rid of large files referenced somewhere in the git history,
> because this will _always_ require a rewrite of the history, correct?

Exactly.
